### PR TITLE
[fix] Correctly creating link anchors in TextEditor & dragging items to hotbar

### LINKED ISSUE #277: FoundryVTT macro bug
### What happened?

Whenever I try to create a macro with a magic item unlocked in the "Items" folder by my DM, it seems to fail. 

### What is the expected behaviour?

The macro should allow me to use the attack without the need to open the character's inventory

### Relevant Errors and Warnings

```shell
![image](https://user-images.githubusercontent.com/120993207/208325580-db9c01b1-4855-42c1-8440-1cd5cf3d45c9.png)
```

### Additional Support Details

_No response_

### Foundry VTT Core Version

V10

### Old-School Essentials Core Version

_No response_

### Please update your browser regularly. Which (updated) browsers are you seeing the problem on?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #278: uuid broken when dragging items from character or monster sheet to chat
### What happened?

1. Drag and drop an item from the character or monster sheet to the chat (or wherever text is displayed)
2. UUID is broken
image

### What is the expected behaviour?

UUID should refer to the item from the item directory

### Relevant Errors and Warnings

_No response_

### Additional Support Details

FVTTT 10 build 291 hosted on The Forge
OSE 1.6.8
Using Brave browser

Active Modules:
----------
Button Feel v;
Cautious Gamemaster's Pack v;
Compendium Folders v;
DF Architect v;
DF Chat Enhancements v;
DF Quality of Life v;
DF Scene Enhancement v;
Dice So Nice! v;
Dice Tray v;
Dig Down - Advanced Search v;
Drag Ruler v;
Dungeon Draw v;
Find the culprit v;
Forien's Quest Log v;
GM Notes v;
Illandril's Chat Enhancements v;
Illandril's Token HUD scaler v;
Inline Table Rolls v;
Item Piles v;
Less Fog v;
lib - Color Settings v;
libWrapper v.0;
LightsHUD v;
Module Compatibility Checker v;
Monk's Little Details v;
Monk's Scene Navigation v;
Monk's Wall Enhancement v;
Multilevel Tokens v;
Multiple Document Selection v;
Ownership Viewer v;
Perfect Vision v;
PnP - Pointer and Pings! v;
PopOut! v;
Quick Insert - Search Widget v;
Simple Calendar v;
SmallTime v;
Smart Target v;
socketlib v;
The Forge v;
The Forge: More Awesomeness v;
Tidy UI - Game Settings v;
Token Auras v;
Token Ease v;
Tokenizer v;
Token Z v;
User Latency v;
Window Controls v;
WTF: What the Flag? v;

### Foundry VTT Core Version

V10

### Old-School Essentials Core Version

1.6.8

### Please update your browser regularly. Which (updated) browsers are you seeing the problem on?

Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by bakbakbakbakbak**
resolves #278

may fix #277 (Depends on if items should be allowed to be dragged from the Items sidebar when not owned by an actor, which is not allowed today)

By using the defined method `toDragData()` from the Item implementation and then adding data necessary, we are able to properly construct the `uuid` which now includes a reference to the `Actor`. While at it I took a look at the macro-code and had to re-write it to reflect changes.

### CODE REVIEW

**Comment by anthonyronda** [comment]
Regarding your comment

> (Depends on if items should be allowed to be dragged from the Items sidebar when not owned by an actor, which is not allowed today)

I'll double check what happens when the behavior here is fixed. There is an override for rolling rollable items (it might just be for weapons) and a default for all items (show item description).

Showing an item description for items you can see somehow is a good overall macro for any helpful player who wants to share the item description for chat.

Allowing a player to roll an item their owned actor doesn't own is confusing. I would say that's not desirable behavior

**Comment by anthonyronda** [comment]
I just did a sanity check in the system

No one, including GM, can drag items from the directory to the macro bar for any reason (even as an "open this item" macro). GM not having this ability is a little unexpected for me but apparently how it has always worked and out of scope for the issue this fixes.

Players with observer-only permission on an actor can't drag items from that actor. Makes sense, likely the core functionality. 

As for the main usecase:

Copying them over now works correctly. 🎉

It appears as though we're not properly error checking in the macros themselves re:checking for a selected actor. If there's not an actor selected, we get console errors instead

```console
Uncaught (in promise) TypeError: actor is undefined
[Detected 2 packages: system:ose-test, lib-wrapper]
    rollItemMacro macros.js:64
    anonymous foundry.js line 20080 > AsyncFunction:3
    _executeScript foundry.js:20082
    execute foundry.js:20037
    _onClickMacro foundry.js:68365
    jQuery 9
    activateListeners foundry.js:68284
    _render foundry.js:5138
    render foundry.js:5067
    initializeUI foundry.js:8071
    setupGame foundry.js:7931
    _initializeGameView foundry.js:8999
    _initializeView foundry.js:8975
    initialize foundry.js:7825
    call_wrapped libWrapper-wrapper.js:507
    libWrapperInit libWrapper-api.js:803
    initialize#0 libWrapper-wrapper.js:187
    <anonymous> foundry.js:79693
    async* foundry.js:79675
```

**Comment by bakbakbakbakbak** [comment]
> 

I'll try to make write a test for it and extend the ticket if I can figure it out!

**Comment by bakbakbakbakbak** [comment]
I discovered that unlinked tokens can't create macros either!

File: src/e2e/actor/createItemMacro.test.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        after(async () => {
+            await trashChat();
+            await trashActor();
+        })
+
+        afterEach(async () => {
+            await trashMacro();
+        })
+
+        it('Create spell macro', async () => { await canCreate("weapon") })
```
Comment (anthonyronda) [inline src/e2e/actor/createItemMacro.test.js:69]: test name should be 'Create weapon macro'

File: src/e2e/actor/createItemMacro.test.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        return Actor.create({
+            ...data,
+            name: testCharacterName,
+            type: 'character'
+        });
+    };
+    const testActor = () => game.actors.getName(testCharacterName);
+    const trashActor = () => testActor()?.delete();
+    const trashMacro = () => (game.macros.size)
+        ? game.macros.documentClass.deleteDocuments([], { deleteAll: true })
```
Comment (anthonyronda) [inline src/e2e/actor/createItemMacro.test.js:25]: we haven't made a test this destructive before. I used to assume non-coders could run tests with Quench themselves, without having to warn them not to use their home worlds.

I'm not sure this needs to be changed but if we can find the testMacro and delete just that, it'd be ideal

**Review by anthonyronda — CHANGES_REQUESTED** [review]
Hi! I ran into some bugs with this one, please take a look

**Comment by anthonyronda** [comment]
This isn't your fault but I just noticed `OSE.warn.noItemWithName` should be `OSE.error.noItemWithName` in `macros.js:67`. While looking over the changes could you fix that please? 😇 🙏

File: src/e2e/actor/createItemMacro.test.js (hunk 2 above)
Comment (bakbakbakbakbak) [reply src/e2e/actor/createItemMacro.test.js:25]: Oh yes, you are 100% right. I'll correct this!

File: src/module/macros.js [hunk 3]
```diff
@@ -11,11 +11,11 @@
  */
 export async function createOseMacro(data, slot) {
   if (data.type !== "Item") return;
-  if (!("data" in data))
+  if (!(!(data.tokenId && data.sceneId) || data.uuid.indexOf("Actor") < 0 ))
```
[the remark this answers is not in the document]
Comment (bakbakbakbakbak) [reply src/module/macros.js]: I thought of this, and there seem to be no possible scenario where an item is dragged from something other than:

1. Actor sheet
2. Items sidebar
3. Compendium

The latest commit checks for the index of `Item.` in the `uuid`-string that is generated for the item using `Item.toDragData()`. Every time this string is generated, the "Item." part of the string is located after the initial position of the string (which was the bug previously, that we didn't use `Item.toDragData()` to generate the string), and the occurences where it is dragged from Item sidebar och a Compendium, it's either in the 0-position or absent from the string.

Thus, the logic was changed to reflect that in commit 4a748cc.

File: src/module/macros.js [hunk 4]
```diff
@@ -11,7 +11,7 @@
  */
 export async function createOseMacro(data, slot) {
   if (data.type !== "Item") return;
-  if (!(!(data.tokenId && data.sceneId) || data.uuid.indexOf("Actor") < 0 ))
+  if ( data.uuid.indexOf("Item.") <= 0 )
```
Comment (anthonyronda) [inline src/module/macros.js:14]: 👏 clever haha

**Review by anthonyronda — APPROVED** [review]
LGTM

File: src/module/macros.js (hunk 4 above)
Comment (bakbakbakbakbak) [reply src/module/macros.js:14]: Sometimes it happens :P

**Pull request merged**